src/msw/dirdlg.cpp

```diff
@@ -378,77 +383,122 @@ int wxDirDialog::ShowIFileDialog(WXHWND owner)
 
-            hr = folder->GetDisplayName(SIGDN_FILESYSPATH, &pathOLE);
-            if ( SUCCEEDED(hr) )
+        hr = itemArray->GetCount(&count);
+        if ( FAILED(hr) )
+        {
+            wxLogApiError(wxS("IShellItemArray::GetCount"), hr);
+            return false;
+        }
+
+        for ( DWORD i = 0; i < count; ++i )
+        {
+            wxCOMPtr<IShellItem> item;
+
+            hr = itemArray->GetItemAt(i, &item);
+            if ( FAILED(hr) )
             {
-                path = pathOLE;
-                CoTaskMemFree(pathOLE);
+                // do not attempt to retrieve any other items
+                // and just fail
+                wxLogApiError(wxS("IShellItemArray::GetItem"), hr);
+                tempPaths.clear();
+                break;
             }
-            else
+
+            if ( !ConvertIShellItemToPath(item, path) )
             {
-                wxLogApiError(wxS("IShellItem::GetDisplayName"), hr);
+                // again, just fail
+                tempPaths.clear();
+                break;
             }
+
+            tempPaths.push_back(path);
         }
-        else
-        {
-            wxLogApiError(wxS("IFileDialog::GetResult"), hr);
-        }
-    }
-    else if ( hr == HRESULT_FROM_WIN32(ERROR_CANCELLED) )
-    {
-        return wxID_CANCEL; // the user cancelled the dialog
+
     }
-    else
+    else // single selection
     {
-        wxLogApiError(wxS("IFileDialog::Show"), hr);
-    }
+        wxCOMPtr<IShellItem> item;
 
-    if ( path.empty() )
-    {
-        // the user didn't cancel the dialog and yet the path is empty
-        // it means there was an error, already logged by wxLogApiError()
-        // now report the error to the user and return
-        wxLogSysError(_("Couldn't obtain folder name"), hr);
-        return wxID_CANCEL;
+        hr = fileDialog->GetResult(&item);
+        if ( FAILED(hr) )
+        {
+            wxLogApiError(wxS("IFileOpenDialog::GetResult"), hr);
+            return false;
+        }
+
+        if ( !ConvertIShellItemToPath(item, path) )
+        {
+            return false;
+        }
+
+        tempPaths.push_back(path);
     }
 
-    m_path = path;
-    return wxID_OK;
+    if ( tempPaths.empty() )
+        return false; // there was en error
+
+    paths = tempPaths;
+    return true;
 }
 
-#endif // wxUSE_IFILEDIALOG
+// helper function for wxDirDialog::ShowIFileOpenDialog()
+bool ConvertIShellItemToPath(const wxCOMPtr<IShellItem>& item, wxString& path)
+{
+    LPOLESTR pathOLE = NULL;
+    const HRESULT hr = item->GetDisplayName(SIGDN_FILESYSPATH, &pathOLE);
 
-// ----------------------------------------------------------------------------
-// private functions
-// ----------------------------------------------------------------------------
+    if ( FAILED(hr) )
+    {
+        wxLogApiError(wxS("IShellItem::GetDisplayName"), hr);
+        return false;
+    }
+
+    path = pathOLE;
+    CoTaskMemFree(pathOLE);
```

Comment:
Not really meant for this PR, but we should have a smart pointer releasing memory using `::CoTaskMemFree()`.

Reply:
I grepped the src folder for `CoTaskMemFree()`.  It is used once by `wxDirDialog`, `wxStandardPaths`, and `wxTaskBarButton`; five times by `wxWebViewEdgeImpl`.

Not sure if this demands creating a new class but I could look into it if you think it would be useful.

Reply:
Well, ideally, yes, I think it would: the rule is to never deallocate any resources manually and always use RAII for them, and this definitely falls under it.

Whether it's worth it compared to the other possible improvements is a different question and, again, not doing it definitely won't prevent this PR from being merged, but it would be nice to not have "raw" calls to free memory.

Reply:
Well, I did add `wxCOMPtr` when adding support for Vista+ style `wxDirDialog` back in 2012, but in this case it can IMO wait till a separate PR. 

ATM, I am not even sure how to name such a smart pointer and what its API should look like.